File: DataEngineering/01_Architecture_and_Storage/projects/golang/04_data_warehouse_schema_builder/main.go
package main

import (
	"fmt"
	"log"
	"time"

	"gorm.io/driver/sqlite"
	"gorm.io/gorm"
	"gorm.io/gorm/logger"
)

// ==========================================
// Schema Definitions (Star Schema)
// ==========================================

// DimDate represents the Date dimension table.
type DimDate struct {
	DateID   uint      `gorm:"primaryKey;autoIncrement"`
	FullDate time.Time `gorm:"uniqueIndex;not null"`
	Year     int       `gorm:"not null"`
	Month    int       `gorm:"not null"`
	Day      int       `gorm:"not null"`
}

// DimProduct represents the Product dimension table.
type DimProduct struct {
	ProductID   uint    `gorm:"primaryKey"`
	ProductName string  `gorm:"size:100;not null"`
	Category    string  `gorm:"size:50;not null"`
	Price       float64 `gorm:"type:decimal(10,2);not null"`
}

// FactSales represents the central fact table.
type FactSales struct {
	SaleID      uint       `gorm:"primaryKey;autoIncrement"`
	DateID      uint       `gorm:"not null"`
	ProductID   uint       `gorm:"not null"`
	Quantity    int        `gorm:"not null"`
	TotalAmount float64    `gorm:"type:decimal(12,2);not null"`
	
	// Foreign Key constraints
	Date    DimDate    `gorm:"foreignKey:DateID"`
	Product DimProduct `gorm:"foreignKey:ProductID"`
}

// ==========================================
// Operations
// ==========================================

func buildAndPopulateWarehouse(db *gorm.DB) error {
	log.Println("Migrating database schema...")
	// AutoMigrate creates tables, missing foreign keys, constraints, columns and indexes.
	err := db.AutoMigrate(&DimDate{}, &DimProduct{}, &FactSales{})
	if err != nil {
		return fmt.Errorf("failed to migrate schema: %w", err)
	}

	log.Println("Inserting sample data into Data Warehouse...")
	
	// Use a transaction for data integrity
	return db.Transaction(func(tx *gorm.DB) error {
		dt := time.Date(2023, 10, 1, 0, 0, 0, 0, time.UTC)
		dDate := DimDate{
			FullDate: dt,
			Year:     dt.Year(),
			Month:    int(dt.Month()),
			Day:      dt.Day(),
		}
		if err := tx.Create(&dDate).Error; err != nil {
			return err
		}

		dProd := DimProduct{
			ProductID:   101,
			ProductName: "Enterprise Database Server",
			Category:    "Hardware",
			Price:       15000.00,
		}
		if err := tx.Create(&dProd).Error; err != nil {
			return err
		}

		fSales := FactSales{
			DateID:      dDate.DateID,
			ProductID:   dProd.ProductID,
			Quantity:    3,
			TotalAmount: dProd.Price * 3,
		}
		if err := tx.Create(&fSales).Error; err != nil {
			return err
		}

		return nil // Commit transaction
	})
}

func analyzeSales(db *gorm.DB) {
	log.Println("Running analytical query on Star Schema...")

	type SalesReport struct {
		FullDate    time.Time
		ProductName string
		Quantity    int
		TotalAmount float64
	}

	var results []SalesReport

	// Perform the joins for our Star Schema query
	err := db.Model(&FactSales{}).
		Select("dim_dates.full_date, dim_products.product_name, fact_sales.quantity, fact_sales.total_amount").
		Joins("JOIN dim_dates ON fact_sales.date_id = dim_dates.date_id").
		Joins("JOIN dim_products ON fact_sales.product_id = dim_products.product_id").
		Scan(&results).Error

	if err != nil {
		log.Fatalf("Failed to execute analytical query: %v", err)
	}

	fmt.Printf("\n--- Sales Analysis Report ---\n")
	fmt.Printf("%-15s | %-30s | %-5s | %s\n", "Date", "Product Name", "Qty", "Total Amount")
	fmt.Println("---------------------------------------------------------------------------")
	for _, row := range results {
		fmt.Printf("%-15s | %-30s | %-5d | $%.2f\n", 
			row.FullDate.Format("2006-01-02"), 
			row.ProductName, 
			row.Quantity, 
			row.TotalAmount)
	}
	fmt.Println("---------------------------------------------------------------------------")
}

func main() {
	// In production, configure connecting to a real data warehouse like PostgreSQL/Redshift/etc.
	// dsn := "host=localhost user=gorm password=gorm dbname=warehouse port=5432 sslmode=disable"
	// db, err := gorm.Open(postgres.Open(dsn), &gorm.Config{})
	
	// For simulation, using SQLite in-memory database
	db, err := gorm.Open(sqlite.Open("file::memory:?cache=shared"), &gorm.Config{
		Logger: logger.Default.LogMode(logger.Warn), // Suppress noisy SQL logs for clear output
	})
	if err != nil {
		log.Fatalf("Failed to connect database: %v", err)
	}

	if err := buildAndPopulateWarehouse(db); err != nil {
		log.Fatalf("Failed to build and populate warehouse: %v", err)
	}

	analyzeSales(db)
}

File: DataEngineering/01_Architecture_and_Storage/projects/python/04_data_warehouse_schema_builder/main.py
import logging
from datetime import datetime
from sqlalchemy import create_engine, Column, Integer, String, Numeric, Date, ForeignKey, select
from sqlalchemy.orm import declarative_base, sessionmaker, relationship

# Setup logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

Base = declarative_base()

# ==========================================
# Schema Definitions (Star Schema)
# ==========================================

class DimDate(Base):
    __tablename__ = 'dim_date'
    
    date_id = Column(Integer, primary_key=True, autoincrement=True)
    full_date = Column(Date, nullable=False, unique=True)
    year = Column(Integer, nullable=False)
    month = Column(Integer, nullable=False)
    day = Column(Integer, nullable=False)

class DimProduct(Base):
    __tablename__ = 'dim_product'
    
    product_id = Column(Integer, primary_key=True)
    product_name = Column(String(100), nullable=False)
    category = Column(String(50), nullable=False)
    price = Column(Numeric(10, 2), nullable=False)

class FactSales(Base):
    __tablename__ = 'fact_sales'
    
    sale_id = Column(Integer, primary_key=True, autoincrement=True)
    date_id = Column(Integer, ForeignKey('dim_date.date_id'), nullable=False)
    product_id = Column(Integer, ForeignKey('dim_product.product_id'), nullable=False)
    quantity = Column(Integer, nullable=False)
    total_amount = Column(Numeric(12, 2), nullable=False)

    # Relationships for easier ORM traversal
    date = relationship("DimDate")
    product = relationship("DimProduct")

# ==========================================
# Operations
# ==========================================

def build_and_populate_warehouse(engine):
    """Creates tables and populates them with initial dimensions and facts."""
    logger.info("Creating database schema...")
    Base.metadata.create_all(engine)
    
    Session = sessionmaker(bind=engine)
    
    with Session() as session:
        try:
            logger.info("Inserting dimension data...")
            # Insert Dimensions
            dt = datetime(2023, 10, 1).date()
            d_date = DimDate(full_date=dt, year=dt.year, month=dt.month, day=dt.day)
            
            d_prod = DimProduct(product_id=101, product_name='Enterprise Database Server', category='Hardware', price=15000.00)
            
            session.add_all([d_date, d_prod])
            session.flush() # Flush to get the auto-generated date_id
            
            logger.info("Inserting fact data...")
            # Insert Facts
            f_sales = FactSales(
                date_id=d_date.date_id, 
                product_id=d_prod.product_id, 
                quantity=3, 
                total_amount=d_prod.price * 3
            )
            session.add(f_sales)
            
            session.commit()
            logger.info("Data committed successfully.")
            
        except Exception as e:
            session.rollback()
            logger.error(f"Error populating data warehouse: {e}")
            raise

def analyze_sales(engine):
    """Demonstrates querying the star schema."""
    logger.info("Running analytical query on Star Schema...")
    Session = sessionmaker(bind=engine)
    
    with Session() as session:
        # Constructing the JOIN query using SQLAlchemy 2.0 select syntax
        stmt = (
            select(
                DimDate.full_date,
                DimProduct.product_name,
                FactSales.quantity,
                FactSales.total_amount
            )
            .select_from(FactSales)
            .join(DimDate, FactSales.date_id == DimDate.date_id)
            .join(DimProduct, FactSales.product_id == DimProduct.product_id)
        )
        
        results = session.execute(stmt).fetchall()
        
        print("\n--- Sales Analysis Report ---")
        print(f"{'Date':<15} | {'Product Name':<30} | {'Qty':<5} | {'Total Amount'}")
        print("-" * 75)
        for row in results:
            print(f"{str(row.full_date):<15} | {row.product_name:<30} | {row.quantity:<5} | ${row.total_amount:,.2f}")
        print("-" * 75)

if __name__ == "__main__":
    # In production, this would be a Postgres/Snowflake/Redshift URI
    # e.g. "postgresql://user:pass@host:5432/warehouse"
    db_uri = "sqlite:///:memory:"
    
    # Create SQLAlchemy engine
    engine = create_engine(db_uri, echo=False)
    
    build_and_populate_warehouse(engine)
    analyze_sales(engine)
